kairos-crypto/src/implementations/casper.rs

```diff
@@ -0,0 +1,128 @@
+use casper_types::bytesrepr::{FromBytes, ToBytes};
+use casper_types::{crypto, PublicKey, SecretKey, Signature};
+use std::path::Path;
+
+use crate::CryptoError;
+use crate::CryptoSigner;
+
+pub struct Signer {
+    private_key: Option<SecretKey>,
+    public_key: PublicKey,
+}
+
+impl CryptoSigner for Signer {
+    fn from_private_key_file<P: AsRef<Path>>(file: P) -> Result<Self, CryptoError>
+    where
+        Self: Sized,
+    {
+        let private_key =
+            SecretKey::from_file(file).map_err(|e| CryptoError::FailedToParseKey {
+                error: e.to_string(),
+            })?;
+        let public_key = PublicKey::from(&private_key);
+
+        Ok(Self {
+            private_key: Some(private_key),
+            public_key,
+        })
+    }
+
+    fn from_public_key(bytes: &[u8]) -> Result<Self, CryptoError>
+    where
+        Self: Sized,
+    {
+        let (public_key, _remainder) =
+            casper_types::PublicKey::from_bytes(bytes).map_err(|_e| {
+                CryptoError::Deserialization {
+                    context: "public key",
+                }
+            })?;
+
+        Ok(Self {
+            private_key: None,
+            public_key,
+        })
+    }
+
+    fn sign(&self, data: &[u8]) -> Result<Vec<u8>, CryptoError> {
+        let private_key = self
+            .private_key
+            .as_ref()
+            .ok_or(CryptoError::MissingPrivateKey)?;
+        let signature = crypto::sign(data, private_key, &self.public_key);
```

Comment:
Mentioned elsewhere, `casper_types::crypto::sign` has an unused `public_key`.

Here is the crypto signing code that `litmus` uses (which has been unit tested and exhaustively tested against `casper_types` for compatibility).

https://github.com/cspr-rad/litmus/blob/673f8182560f7f1ff6fee04d775070a6b3e380c3/src/crypto.rs#L24-L40

```rust
// Signs the given message using the given key.
pub fn sign<T: AsRef<[u8]>>(secret_key: &SecretKey, message: T) -> Signature {
    match secret_key {
        SecretKey::System => Signature::System,
        SecretKey::Ed25519(secret_key) => {
            let signature = secret_key.sign(message.as_ref());
            Signature::Ed25519(signature)
        }
        SecretKey::Secp256k1(secret_key) => {
            let signature = secret_key
                .try_sign(message.as_ref())
                .expect("should create signature");
            Signature::Secp256k1(signature)
        }
        _ => panic!("SecretKey is marked as non-exhaustive, but this should never happen"),
    }
}
```

Reply:
I am aware that public key is not used internally, but I see no point in duplicating the Casper implementation just to get rid of it:

![image](https://github.com/cspr-rad/kairos/assets/6506780/3ec2fe00-3b39-4167-b2d1-90f810e56691)

Additionally, public key is always available in `Signer`, so there is no extra code necessary.